File: packages/types/src/agents.ts
// ─── Agents explorer ─────────────────────────────────────────────
// Definitions and settings mirror `tracepilot_orchestrator::agents`;
// usage mirrors `tracepilot_core::analytics::agents`.

/** Where a definition comes from. */
export type AgentScope = "builtin" | "personal" | "project" | "plugin";

/** Built-in `*.agent.yaml`, or `*.agent.md` with YAML frontmatter. */
export type AgentFormat = "yaml" | "markdown";

/**
 * Editable definition fields. `null` means the key is absent; a save only
 * rewrites keys that changed, preserving comments and unknown keys.
 */
export interface AgentFields {
  name: string | null;
  displayName: string | null;
  description: string | null;
  /** Ordered model preferences (a single `model:` string, or fallbacks). */
  models: string[];
  /** `model-policy`, e.g. `required` (never fall back). */
  modelPolicy: string | null;
  reasoningEffort: string | null;
  contextTier: string | null;
  /** `null` = key absent: the agent gets every tool. */
  tools: string[] | null;
  includeCustomInstructions: boolean | null;
  deferredToolLoading: boolean | null;
  disableModelInvocation: boolean | null;
  userInvocable: boolean | null;
  infer: boolean | null;
}

export interface AgentOtherField {
  key: string;
  value: string;
}

export interface AgentDiagnostic {
  path: string;
  message: string;
  severity: "error" | "warning" | "info";
}

export interface AgentDefinitionSummary {
  /** The file path. */
  id: string;
  /** Name the CLI dispatches by (frontmatter `name`, else the file stem). */
  name: string;
  fileStem: string;
  displayName: string | null;
  description: string;
  scope: AgentScope;
  format: AgentFormat;
  path: string;
  sourceLabel: string;
  repoRoot: string | null;
  fields: AgentFields;
  hasMcpServers: boolean;
  readOnlyReason: string | null;
  modifiedAt: string | null;
}

export interface AgentDefinitionDetail {
  summary: AgentDefinitionSummary;
  rawContent: string;
  /** The prompt: Markdown body or the YAML `prompt` value. */
  body: string;
  otherFields: AgentOtherField[];
  mcpServers: unknown | null;
  diagnostics: AgentDiagnostic[];
}

/** `subagents.agents.<agentType>` in settings.json (`/subagents`). */
export interface SubagentOverride {
  /** A model ID or `inherit` (the session model). */
  model: string | null;
  effortLevel: string | null;
  /** `inherit` | `default` | `long_context` */
  contextTier: string | null;
  /** Other keys kept as-is (e.g. `autoInvoke`). */
  otherKeys?: string[];
}

export interface SubagentSettings {
  settingsPath: string;
  overrides: Record<string, SubagentOverride>;
  disabled: string[];
  maxConcurrency: number | null;
  maxDepth: number | null;
  sessionModel: string | null;
  sessionEffort: string | null;
  /** Set when settings.json has an unexpected shape: overrides are read-only. */
  shapeError: string | null;
  raw: unknown | null;
}

export interface AgentCatalog {
  definitions: AgentDefinitionSummary[];
  diagnostics: AgentDiagnostic[];
  settings: SubagentSettings;
  /** CLI version whose built-in definitions were read. */
  cliVersion: string | null;
  personalDir: string;
  repoRoots: string[];
}

export type AgentCreateScope = "personal" | "project";

export interface AgentWriteResult {
  path: string;
  backupPath: string | null;
}

// ─── Usage ───────────────────────────────────────────────────────

/** Percentiles over runs that reported a value; `count` is the denominator. */
export interface MetricDistribution {
  count: number;
  min: number | null;
  p25: number | null;
  p50: number | null;
  p75: number | null;
  p90: number | null;
  max: number | null;
}

export interface AgentLabelCount {
  label: string;
  runs: number;
}

export interface AgentBucketCount {
  value: number;
  runs: number;
}

export interface AgentDayCount {
  date: string;
  runs: number;
}

export interface AgentUsageStats {
  name: string;
  agentType: string | null;
  displayName: string | null;
  description: string | null;
  runs: number;
  sessions: number;
  completed: number;
  failed: number;
  cancelled: number;
  incomplete: number;
  durationMs: MetricDistribution;
  /** Can include descendants: never sum across a hierarchy. */
  totalTokens: MetricDistribution;
  toolCalls: MetricDistribution;
  /** Median duration over the equally long window before the range. */
  previousMedianDurationMs: number | null;
  runsWithCredits: number;
  ownNanoAiu: number | null;
  firstUsed: string | null;
  lastUsed: string | null;
  topModels: AgentLabelCount[];
  mismatchRuns: number;
  runsWithConfiguration: number;
  maxDepth: number;
  peakSiblings: number;
  followUps: number;
  multiTurnRuns: number;
  dailyRuns: AgentDayCount[];
}

export interface AgentSelectionStats {
  name: string;
  displayName: string | null;
  sessions: number;
  lastSelected: string | null;
}

export interface AgentUsageSummary {
  totalRuns: number;
  totalSessions: number;
  failedRuns: number;
  cancelledRuns: number;
  incompleteRuns: number;
  maxDepth: number;
  peakParallelism: number;
  runsWithCredits: number;
  totalOwnNanoAiu: number;
  agents: AgentUsageStats[];
  mainAgentSelections: AgentSelectionStats[];
}

export interface AgentDayOutcomes {
  date: string;
  completed: number;
  failed: number;
  cancelled: number;
  incomplete: number;
}

export interface AgentDispatchCount {
  configuredModel: string | null;
  firstDispatchedModel: string | null;
  actualModel: string | null;
  overrideReason: string | null;
  runs: number;
}

export interface AgentParentCount {
  /** `null` for the main agent. */
  parent: string | null;
  runs: number;
}

export interface AgentFailureReason {
  reason: string;
  example: string;
  runs: number;
  lastSeen: string | null;
}

export type AgentRunOutcome = "completed" | "failed" | "cancelled" | "incomplete";

export interface AgentRunRecord {
  sessionId: string;
  sessionSummary: string | null;
  repository: string | null;
  runKey: string;
  toolCallId: string | null;
  displayName: string | null;
  description: string | null;
  startedAt: string | null;
  outcome: AgentRunOutcome;
  errorText: string | null;
  model: string | null;
  durationMs: number | null;
  totalTokens: number | null;
  totalToolCalls: number | null;
  ownNanoAiu: number | null;
  depth: number;
  parentAgentName: string | null;
  turnIndex: number;
  eventIndex: number | null;
}

export interface AgentUsageDetail {
  stats: AgentUsageStats;
  outcomesByDay: AgentDayOutcomes[];
  dispatch: AgentDispatchCount[];
  invokedBy: AgentParentCount[];
  depths: AgentBucketCount[];
  parallelism: AgentBucketCount[];
  failureReasons: AgentFailureReason[];
  executionModes: AgentLabelCount[];
  repositories: AgentLabelCount[];
  recentRuns: AgentRunRecord[];
}

File: crates/tracepilot-core/src/analytics/agents.rs
//! Cross-session agent usage types served to the Agents explorer.
//!
//! Token figures come from `subagent.completed.totalTokens`, which can include
//! descendants: they describe one run and are never summed across a
//! hierarchy. Credits come only from the `agentMetrics` ledger, so every
//! credit total carries the number of runs it covers.

use serde::{Deserialize, Serialize};

/// Percentiles over the runs that reported a value. `count` is the
/// denominator: runs without the metric (older CLIs) are excluded.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MetricDistribution {
    pub count: u64,
    pub min: Option<u64>,
    pub p25: Option<u64>,
    pub p50: Option<u64>,
    pub p75: Option<u64>,
    pub p90: Option<u64>,
    pub max: Option<u64>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentLabelCount {
    pub label: String,
    pub runs: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentBucketCount {
    pub value: u32,
    pub runs: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentDayCount {
    pub date: String,
    pub runs: u64,
}

/// Aggregate usage for one agent name (case-insensitive).
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentUsageStats {
    /// Most recently seen spelling of the agent name.
    pub name: String,
    pub agent_type: Option<String>,
    /// Display name / description, only when every run in range reported the
    /// same value: per-call names and descriptions describe a run, not the agent.
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub runs: u64,
    pub sessions: u64,
    pub completed: u64,
    pub failed: u64,
    pub cancelled: u64,
    pub incomplete: u64,
    pub duration_ms: MetricDistribution,
    /// Includes descendants where the CLI reports it that way.
    pub total_tokens: MetricDistribution,
    pub tool_calls: MetricDistribution,
    /// Median duration over the equally long window before the range, for
    /// trend-based slowness. `None` without a bounded range or earlier runs.
    pub previous_median_duration_ms: Option<u64>,
    /// Runs covered by the credit ledger (1.0.83+) and their exclusive total.
    pub runs_with_credits: u64,
    pub own_nano_aiu: Option<u64>,
    pub first_used: Option<String>,
    pub last_used: Option<String>,
    pub top_models: Vec<AgentLabelCount>,
    /// Runs whose dispatched or actual model differed from the configured one.
    pub mismatch_runs: u64,
    /// Runs recording a configured model (the mismatch denominator).
    pub runs_with_configuration: u64,
    pub max_depth: u32,
    pub peak_siblings: u32,
    pub follow_ups: u64,
    pub multi_turn_runs: u64,
    pub daily_runs: Vec<AgentDayCount>,
}

/// Sessions whose main agent was a selected custom agent (`--agent`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentSelectionStats {
    pub name: String,
    pub display_name: Option<String>,
    pub sessions: u64,
    pub last_selected: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentUsageSummary {
    pub total_runs: u64,
    pub total_sessions: u64,
    pub failed_runs: u64,
    pub cancelled_runs: u64,
    pub incomplete_runs: u64,
    pub max_depth: u32,
    pub peak_parallelism: u32,
    pub runs_with_credits: u64,
    pub total_own_nano_aiu: u64,
    pub agents: Vec<AgentUsageStats>,
    pub main_agent_selections: Vec<AgentSelectionStats>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentDayOutcomes {
    pub date: String,
    pub completed: u64,
    pub failed: u64,
    pub cancelled: u64,
    pub incomplete: u64,
}

/// Configured vs. dispatched vs. actual model for 1.0.83+ runs.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentDispatchCount {
    pub configured_model: Option<String>,
    pub first_dispatched_model: Option<String>,
    pub actual_model: Option<String>,
    pub override_reason: Option<String>,
    pub runs: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentParentCount {
    /// `None` for the main agent.
    pub parent: Option<String>,
    pub runs: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentFailureReason {
    /// Normalised error text (IDs and numbers masked).
    pub reason: String,
    pub example: String,
    pub runs: u64,
    pub last_seen: Option<String>,
}

/// One run, for the Recent runs list and deep links into the session.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentRunRecord {
    pub session_id: String,
    pub session_summary: Option<String>,
    pub repository: Option<String>,
    pub run_key: String,
    pub tool_call_id: Option<String>,
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub started_at: Option<String>,
    pub outcome: String,
    pub error_text: Option<String>,
    pub model: Option<String>,
    pub duration_ms: Option<u64>,
    pub total_tokens: Option<u64>,
    pub total_tool_calls: Option<u64>,
    pub own_nano_aiu: Option<u64>,
    pub depth: u32,
    pub parent_agent_name: Option<String>,
    pub turn_index: u64,
    pub event_index: Option<u64>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentUsageDetail {
    pub stats: AgentUsageStats,
    pub outcomes_by_day: Vec<AgentDayOutcomes>,
    pub dispatch: Vec<AgentDispatchCount>,
    pub invoked_by: Vec<AgentParentCount>,
    pub depths: Vec<AgentBucketCount>,
    pub parallelism: Vec<AgentBucketCount>,
    pub failure_reasons: Vec<AgentFailureReason>,
    pub execution_modes: Vec<AgentLabelCount>,
    pub repositories: Vec<AgentLabelCount>,
    pub recent_runs: Vec<AgentRunRecord>,
}
